src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java

```diff
@@ -139,6 +141,8 @@ public class LettuceConnectionFactory implements RedisConnectionFactory, Reactiv
 	/** Synchronization monitor for the shared Connection */
 	private final Object connectionMonitor = new Object();
```

Comment:
@mp911de 
I expected `connectionMonitor` field and other `synchronized` blocks (e.g in getOrCreateSharedConnection() method) to be deleted, but it was not.
Is it ok?

Reply:
`connectionMonitor` is used in `synchronized` blocks that are of a computational nature to enforce happens-before relationships.

Reply:
@mp911de 
Ok, I see.

Reply:
The `connectionMonitor` is only really used to initialize references to the single, "shared" connection(s) [here](https://github.com/spring-projects/spring-data-redis/blob/3.2.0-M2/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java#L1140-L1147) and [here](https://github.com/spring-projects/spring-data-redis/blob/3.2.0-M2/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java#L1152-L1159), then subsequently (and necessarily) [used](https://github.com/spring-projects/spring-data-redis/blob/3.2.0-M2/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java#L1120-L1124) to uninitialize the references once the connections are reset (released).

There is nothing "computationally" happening within these `synchronized` blocks, AFAICT. 

For instance, the connection is not actually established yet. That only occurs further up in the call stack, inside `getSharedConnection(..)` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.2.0-M2/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java#L1175)) and `getSharedReactiveConnection(..)` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.2.0-M2/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java#L1200)), which do not get executed inside a `synchronized` block on the `connectionMonitor`.

Unless I am overlooking something (@mp911de - correct me if I am wrong), technically, it should not matter what (Virtual or Platform) Thread initializes (gets or creates) and resets the connections, given that all the `connectionMonitor` is used for is to un/initialize references to the connections, thereby ensuring only a **single**, "shared" connection (for both reactive and non-reactive) is created and returned to the calling Thread.

> NOTE: `ReentrantLock` has the same mutex, visibility and reentrant behavior as `synchronized` blocks (see [Javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/locks/ReentrantLock.html)).

If by "_happens-before_" you mean the `null` check before, or the return statement after the creation of the connection, then logically the statement ordering as seen by the Thread and determined by the runtime will necessarily be as it appears in code. Semantically speaking, it has to (collectively, the creation is inside the "`if`" statement so is a moot point; it only boils down to the `if` and `return` statements, which semantically cannot be reordered by the runtime in this case). This is `true` regardless if the code is inside a `synchronized` block (or even `ReentrantLock`) or not.

Again, maybe I am overlooking something here?

Reply:
I imagine there should be some benefit here in different Virtual Threads acquiring the "shared" connection from the factory in the case of a highly concurrent Spring-based Web application, where Virtual Threads are configured. Underneath, the user maybe using a pooled `LettuceConnectionProvider`.

Thoughts?

Reply:
After discussing this with @mp911de this morning, I made the change to the `synchronized` blocks in `LettuceConnectionFactory` to now use the `ReentrantLock`.